v2/workloadapi/client.go

```diff
@@ -2,75 +2,411 @@ package workloadapi
+		return nil, err
+	}
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, err
+	}
+	result := []*x509svid.SVID{}
+	for _, svid := range resp.Svids {
+		s, err := x509svid.ParseRaw(svid.X509Svid, svid.X509SvidKey)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, s)
+	}
+	return result, nil
 }
 
 // FetchX509Bundle fetches the X.509 bundles.
 func (c *Client) FetchX509Bundles(ctx context.Context) (*x509bundle.Set, error) {
-	panic("not implemented")
+	ctx, cancel := context.WithCancel(withHeader(ctx))
+	defer cancel()
+	stream, err := c.wlClient.FetchX509SVID(ctx, &workload.X509SVIDRequest{})
+	if err != nil {
+		return nil, err
+	}
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, err
+	}
+	bundles := []*x509bundle.Bundle{}
+	for _, svid := range resp.Svids {
+		td, err := spiffeid.TrustDomainFromString(svid.SpiffeId)
+		if err != nil {
+			return nil, err
+		}
+		certs, err := x509.ParseCertificates(svid.Bundle)
+		if err != nil {
+			return nil, err
+		}
+		b := x509bundle.FromX509Roots(td, certs)
+		bundles = append(bundles, b)
+	}
+	return x509bundle.NewSet(bundles...), nil
 }
 
 // FetchX509Context fetches the X.509 context, which contains both X509-SVIDs
 // and X.509 bundles.
 func (c *Client) FetchX509Context(ctx context.Context) (*X509Context, error) {
-	panic("not implemented")
+	ctx, cancel := context.WithCancel(withHeader(ctx))
+	defer cancel()
+	stream, err := c.wlClient.FetchX509SVID(ctx, &workload.X509SVIDRequest{})
+	if err != nil {
+		return nil, err
+	}
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, err
+	}
+	svids := []*x509svid.SVID{}
+	bundles := []*x509bundle.Bundle{}
+	for _, svid := range resp.Svids {
+		s, err := x509svid.ParseRaw(svid.X509Svid, svid.X509SvidKey)
+		if err != nil {
+			return nil, err
+		}
+		svids = append(svids, s)
+
+		certs, err := x509.ParseCertificates(svid.Bundle)
+		if err != nil {
+			return nil, err
+		}
+		b := x509bundle.FromX509Roots(s.ID.TrustDomain(), certs)
+		bundles = append(bundles, b)
+	}
+	return &X509Context{
+		SVIDs:   svids,
+		Bundles: x509bundle.NewSet(bundles...),
+	}, nil
 }
 
 // WatchX509Context watches for updates to the X.509 context. The watcher
 // receives the updated X.509 context.
 func (c *Client) WatchX509Context(ctx context.Context, watcher X509ContextWatcher) error {
-	panic("not implemented")
+	for {
+		err := c.watchX509Context(ctx, watcher)
+		watcher.OnX509ContextWatchError(err)
+		err = c.handleCallError(ctx, err)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (c *Client) handleCallError(ctx context.Context, err error) error {
+	code := status.Code(err)
+	if code == codes.Canceled || code == codes.InvalidArgument {
+		c.opts.log.Errorf("cancelling : %v", err)
+		return err
+	}
+	c.opts.log.Errorf("there was an error with the gRPC call: %v", err)
+	c.opts.log.Infof("backing off before retrying")
+
+	select {
+	case <-time.After(c.backoff.Duration()):
+		c.opts.log.Infof("retrying")
```

Comment:
```suggestion
	c.opts.log.Errorf("Failed to watch the Workload API: %v", err)
	backoff := c.backoff.Duration()
	c.opts.log.Debugf("Retrying watch in %s", backoff)

	select {
	case <-time.After(backoff):
```

Reply:
Instead of logging "retrying" here, maybe that should be a debug log line at the top of the `for {` loop on each watch. That way it can be specific.

```
c.opts.log.Debugf("Watching X.509 contexts")
```